File: Array/LC_271_Encode_and_Decode_Strings.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.17

class Codec {
public:

    // Encodes a list of strings to a single string.
    string encode(vector<string>& strs) {
        // use number + "#" to indicate the length of single str
        // ex ["Hello", "World"] -> 5#Hello5#World
        string result = "";
        for (string str: strs){
            string size = to_string(str.size());
            result += size + "#" + str;
        }
        return result;
    }

    // Decodes a single string to a list of strings.
    vector<string> decode(string s) {
        vector<string> result;
        int i = 0;
        while (i < s.size()){
            int j = i;
            while (s[j] != '#'){
                j++;
            }
            // substr(start, length) --> equal to s[start: start + length]
            int size = stoi(s.substr(i, j - i));
            result.push_back(s.substr(j+1, size));
            i = j + 1 + size;
        }
        return result;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.decode(codec.encode(strs));

File: Array/LC_271_Encode_and_Decode_Strings.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.17

class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        # give string length and "#" as seperation
        # ["Hello", "World"] -> 5#Hello5#World
        result = ""
        for string in strs:
            result += str(len(string)) + "#" + string
        return result
            
        

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        ans = []
        i = 0
        while i < len(s):
            j = i
            while j < len(s) and s[j] != "#":
                j += 1
            lens = int(s[i:j])
            ans.append(s[j+1:j+1+lens]) # word starts from j+1, since str[j] is #
            i = j+1+lens
        return ans


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))